"What's New" credit and referencing

trexfeathers opened the pull request:
Adds two new aspects to Iris "What's New" entries to improve Iris' efficacy as an open source project.
* **Pull Request references**
Open source projects need to be both dynamic and transparent, so it's important that anyone interested can get as much detail as they need on changes in the project.
* **Crediting developers**
Open source communities rely on their developer base, some of whom might be volunteering their time. Public recognition rewards contributing developers and demonstrates the quality of the community, hopefully encouraging further and wider collaboration.

Mimicking the [Cartopy format](https://scitools.org.uk/cartopy/docs/latest/whats_new.html) - developer credit first - meant adapting the 'voice' of many of the existing "What's New" entries to get them to state the implemented change before any further context. If anyone below feels I have minced their words, **or if they are unhappy being publicly credited**, feel free to speak up!
* @MoseleyS 
* @stephenworsley 
* @pp-mo 
* @abooton 
* @bouweandela 
* @bjlittle 
* @trexfeathers 
* @jonseddon 
* @tkknight

trexfeathers commented:
Closes #3818

bouweandela commented on docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:
```diff
@@ -23,27 +23,41 @@ contain multiple changes that are worth highlighting as contributions to the
-  * Fixed :issue:`9999`.  Lorem ipsum dolor sit amet, consectetur adipiscing 
-    elit, sed do eiusmod tempor incididunt ut labore et dolore magna aliqua. 
-    <blank line>
+* Names of those who contributed the change e.g. ``Robert FitzRoy and Luke
```
An alternative way of doing this could be to use the names contributors use publicly on GitHub (and link to their profile page), as done e.g. here: https://docs.dask.org/en/latest/changelog.html#id16. That has the advantage that it's up to contributors themselves to decide if they want to be listed with their real name.

trexfeathers replied:
Thanks @bouweandela , I had considered this but I only checked Cartopy, SciPy and Numpy so I hadn't seen any precedent. Good to know that Dask do it - I will make the change

trexfeathers replied:
596f31b7aa678cecb4903b0898cbbeef73aba594

trexfeathers commented:
@bouweandela as one of our newest core devs do you fancy reviewing and merging your first Iris PR?

trexfeathers commented:
Ported over the contents of #3843 - felt like I was asking for Git trouble otherwise!

Closes #3837 

Adds:
* More What's New entries for important pre- 4761a10630b044aedb38f56f96ecff3f52085ca4 work that had been missed
  Contributions from @stephenworsley , @trexfeathers , @bjlittle , @pp-mo , @lbdreyer (in addition to credits in #3838 )

---

* Consistent "What's New" ordering and headings between docs, template and `latest.rst`
  With _Internal_ now at the bottom
* Updated docs on what is expected regarding "What's New" entries
* Updated docs pull request checklist to bring in line with current verbally agreed practices
* Added pull request template referencing the docs pull request checklist

bjlittle commented on .github/pull_request_template.md:
```diff
@@ -0,0 +1,5 @@
+as possible with the
+[Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/pulls.html#the-iris-check-list).
+
+---
```
@trexfeathers Nice! I was thinking along the lines of something like this...


## 🚀 Pull Request

### Description
<!-- Provide a clear description about your awesome pull request -->
<!-- Tell us all about your new feature, improvement, or bug fix -->



### Additional information
<!-- Provide any further information to help the reviewer understand -->
<details>
<summary>Click to expand <b>this section...</b></summary>

```
Add additional verbose information in a collapsible section.
```
See [here](https://gist.github.com/pierrejoubert73/902cc94d79424356a8d20be2b382e1ab) for further details.
</details>

---
[Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/pulls.html#the-iris-check-list)

bjlittle replied:
What do you think?

trexfeathers replied:
I'm not sure what problems this solves. We don't have PR categories so the title is purely aesthetic, and it's pretty rare for a PR to be so detailed that it needs headings/collapsing. Indeed those 5% come from devs who probably know these features already. 

I'm worried we're just going to encourage a load of markdown 'artifacts' on the 95% of the PR's that are small enough to be better without them. Are you keen for/expecting PR's to get more detailed in future? 

Now I've seen it I quite like the idea of _only_ including the checklist link in the way you have written it.

bjlittle replied:
I personally find this pretty minimal myself. If anything you could drop the `Additional information` section... I think you're right, it would most likely be "just deleted" by most PR authors.

However, what I am keen that we do is have is a consistent branding on GitHub for `iris`, and as such this template suggestion is pretty much in line with the new [Issue templates](https://github.com/SciTools/iris/issues/new/choose).

So, we could trim it down to simply...

## 🚀 Pull Request

### Description
<!-- Provide a clear description about your awesome pull request -->
<!-- Tell us all about your new feature, improvement, or bug fix -->


---
[Iris pull request check list]( https://scitools-iris.readthedocs.io/en/latest/developers_guide/pulls.html#the-iris-check-list)

bjlittle replied:
BTW I think that the actual [Iris pull request check list](https://scitools-iris.readthedocs.io/en/latest/developers_guide/pulls.html#the-iris-check-list) needs a bit of work... but @tkknight may have that in hand :thinking:

trexfeathers replied:
If consistency is the key factor then I appreciate where you're coming from. I'll make the change either today or Monday

> BTW I think that the actual Iris pull request check list needs a bit of work...

Is that even after [the changes to the checklist that this PR suggests](https://github.com/SciTools/iris/pull/3838/commits/15ebb9965d6552900b7fd881bfcb3b729681dd5e#diff-8cd7796992d8299a5dbab9c6080157eb)?

bjlittle replied:
Oops, sorry... Hasn't got that far 🤭

tkknight replied:
FYI, the **developers guide** overhaul now has a branch in my forked iris repo.  It still needs some work.  The PR checklist has been improved I can **rebase** if it is updated in @trexfeathers pr.

- [My branch](https://github.com/SciTools/iris/compare/master...tkknight:developer_guide_overhaul)
- [My branch pr checklist](https://github.com/SciTools/iris/compare/master...tkknight:developer_guide_overhaul#diff-107170be3096da8087bc84d4a3018b83)

trexfeathers replied:
@bjlittle 08ee8ba8fb32244e7c566138dfae0b7beb4f801d

bouweandela replied:
Maybe "Description" could be replaced by "What's new" and the content of that section could then be used to compose the complete "What's new" file when making a new release by combining all the entries from merged pull requests since the previous release, as also proposed here https://github.com/SciTools/iris/issues/3837#issuecomment-691956633? To keep it easy to use, harvesting entries and combining them into a latest.rst file would then be done by a script using the GitHub API.

trexfeathers replied:
Totally 👍 on this (https://github.com/SciTools/iris/issues/3837#issuecomment-691980738), but I'm very reluctant to introduce something new to the docs build at such a late stage for v3.0.0

bouweandela replied:
It's up to you, you know your own timeline best.

rcomer commented on docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:
```diff
@@ -10,8 +10,34 @@ document are written by the developer most familiar with the change made.
+code is ready for review, then create a **new pull request** specifically for
+the "What's New" entry, referencing the main pull request and titled
+(e.g. for PR#9999):
+
```
Sometimes the review can take a while, e.g. because no reviewers are available or because the change turns out to be more complicated than the developer realised (the latter happens to me a lot!)  I wonder if it would be better to leave the whatsnew PR until the main one is almost done.

trexfeathers replied:
I agree this is a problem. But it would be ideal if the reviewer would still be able to insist on a "What's New" before merging. Do you think it would work for the timeline to look something like: "OK I'm now happy we've sorted the code changes, please can you create a "What's New" PR and we'll be done."?

rcomer replied:
That seems reasonable to me.  Having said that, if the main PR is almost ready to merge, it might be simpler to rebase and add the whatsnew to the existing branch.

Rebase would also ensure that Travis is looking at the right thing.  I did once manage to [break master with a failure that didn’t show up in my PR](https://github.com/SciTools/iris/pull/2960#issuecomment-385633071). 😬

trexfeathers replied:
>Rebase would also ensure that Travis is looking at the right thing. I did once manage to break master with a failure that didn’t show up in my PR. 😬

As I understand, Travis simulates a merge with the default branch, then runs the tests. So rather than rebasing I **think** the problem you reference would have been caught by re-spinning Travis just before merging the PR.

trexfeathers replied:
8158666c0a8432af262e9141c7167975501a61b4

rcomer commented on docs/iris/src/whatsnew/latest.rst:
```diff
@@ -13,193 +13,255 @@ This document explains the changes made to Iris for this release
   to format date/time coordinates for use on a plot axis (previously
   :meth:`~iris.plot.pcolor` and :meth:`~iris.plot.pcolormesh` did not include
-  this behaviour).
+  this behaviour). (:pull:`3762`)
```
I haven’t checked, but it looks like this fixed #3347.

rcomer replied:
I've now checked with `master` and #3347 is indeed fixed :sunglasses:.

trexfeathers replied:
Wow, the "What's New" is working even before it gets merged 😂

trexfeathers commented on docs/iris/src/developers_guide/pulls.rst:
```diff
@@ -1,7 +1,7 @@
 .. _pr_check:
```
Changes proposed to this file are the minimum to bring the check list up to date. I expect them to be soon over-written by @tkknight 's wholesale re-write of the Developer's Guide - [see this comment](https://github.com/SciTools/iris/pull/3838#discussion_r487403532)

stephenworsley commented on docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:
```diff
@@ -22,28 +56,49 @@ which improved Iris in some way. As such, a single Iris Pull Request may
+
+
+The above example also demonstrates some of the possible syntax for including
+links to code. For more inspiration on possible content and references please
```
There should be a comma after "references".

trexfeathers replied:
3138a01bbee3e9e2c67644ff9b4eae4eddde73aa

stephenworsley commented on docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:
```diff
@@ -64,12 +119,15 @@ users.  To achieve this several categories may be used.
+
+*Internal*
+  Changes to any internal or development related topics, such as testing,
+  environment dependencies etc
```
There should be a full stop after "etc".

trexfeathers replied:
3138a01bbee3e9e2c67644ff9b4eae4eddde73aa

stephenworsley commented on docs/iris/src/whatsnew/latest.rst:
```diff
@@ -13,193 +13,255 @@ This document explains the changes made to Iris for this release
-* The method :meth:`~iris.Cube.cube.remove_coord` would fail to remove derived
-  coordinates, will now remove derived coordinates by removing aux_factories.
+* `stephenworsley`_ fixed :meth:`~iris.Cube.cube.remove_coord` to now also
+  remove derived coordinates, by removing aux_factories. (:pull:`3641`)
```
The comma here should be removed.

trexfeathers replied:
d1038c403d23de9e461e3fe5e3a94cb7def73778

stephenworsley commented on docs/iris/src/whatsnew/latest.rst:
```diff
@@ -13,193 +13,255 @@ This document explains the changes made to Iris for this release
   :meth:`~iris.cube.CubeList.extract_cubes`.
   The new routines perform the same operation, but in a style more like other
-  ``Iris`` functions such as :meth:`~iris.load_cube` and :meth:`~iris.load_cubes`.
+  ``Iris`` functions such as :meth:`~iris.load_cube` and:meth:`~iris.load_cubes`.
```
There should be a space between "and" and ":meth:"

trexfeathers replied:
d1038c403d23de9e461e3fe5e3a94cb7def73778

stephenworsley commented on docs/iris/src/whatsnew/latest.rst:
```diff
@@ -13,193 +13,255 @@ This document explains the changes made to Iris for this release
-* Change tests to account for more detailed descriptions of projections in
-  `GDAL <https://github.com/OSGeo/gdal>`_ 
-  (`see GDAL#1185 <https://github.com/OSGeo/gdal/pull/1185>`_).
+* `stephenworsley`_ changed :attr:`iris.fileformats.pp.PPField.lbproc` is be an
```
"is be an" should be "to be an"

trexfeathers replied:
d1038c403d23de9e461e3fe5e3a94cb7def73778

stephenworsley commented on docs/iris/src/whatsnew/latest.rst:
```diff
@@ -13,193 +13,255 @@ This document explains the changes made to Iris for this release
+  support. (:pull:`3513`)
 
+* `Bill Little`_ improved the developer set up process. Configuring Iris and
+  :ref:`installing_from_source` as a developer, with all the required package
```
Comma should probably be removed

trexfeathers replied:
d1038c403d23de9e461e3fe5e3a94cb7def73778

stephenworsley reviewed (changes requested):
Mostly just a couple typos to fix. Otherwise this is looking good. Good work!

trexfeathers commented on docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:
```diff
@@ -10,8 +10,42 @@ document are written by the developer most familiar with the change made.
+
+* PR reviewer: once the "What's New" PR is created, **merge the main PR**.
+  (If any links in the "What's New" PR referenced new features, these should
+  now be working).
```
Good point @stephenworsley . This workflow is an effort to avoid needing to rebase, and there's every chance that a rebase to the "What's New" PR might be followed by actual "What's New" corrections and it would all get pushed up.

I don't see an easy solution, and in the long term I'm personally in favour of @bouweandela 's proposal (https://github.com/SciTools/iris/issues/3837#issuecomment-691956633), so in the meantime do you think it would be OK to just make specific reference to Travis' linkcheck test as being the thing that `should now be working`?

stephenworsley replied:
That sounds fair to me.

trexfeathers replied:
544d324c4cc638fbccc06cbcde9e05a170a1a205

trexfeathers commented on docs/iris/src/developers_guide/documenting/whats_new_contributions.rst:
```diff
@@ -10,8 +10,42 @@ document are written by the developer most familiar with the change made.
+=============
+
+If changes to ``latest.rst`` are being suggested in several simultaneous
+Iris Pull Requests, Git will likely encounter merge conflicts. In this
```
3138a01bbee3e9e2c67644ff9b4eae4eddde73aa

trexfeathers commented:
Will need a follow up PR to correct the ["What's New" from #3804 ](https://github.com/SciTools/iris/pull/3804/files#diff-7f35cedb5073b58582e2a3fa56496ca2R95-R99)

pp-mo commented:
I see that you are mostly using full real names, but not for everyone, e.g [here](https://github.com/SciTools/iris/pull/3838/files#diff-7f35cedb5073b58582e2a3fa56496ca2R266) and [here](https://github.com/SciTools/iris/pull/3838/files#diff-7f35cedb5073b58582e2a3fa56496ca2R217).
I assume you have based this on the public user account details, hence e.g. @lbdreyer is not identified.  
I guess that is ok + means you don't have to ask everyone what they are happy with.

However. I think it might look better, and be safer in terms of what contributors are really comfortable with, to **just use user ids**.  
Certainly, I always stick to userids in discussions, where I don't want to give any appearance of relying on offline relations.

trexfeathers commented:
> I see that you are mostly using full real names, but not for everyone, e.g here and here.
I assume you have based this on the public user account details, hence e.g. @lbdreyer is not identified.
I guess that is ok + means you don't have to ask everyone what they are happy with.

https://github.com/SciTools/iris/pull/3838#discussion_r486827236
